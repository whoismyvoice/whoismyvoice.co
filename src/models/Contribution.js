// @flow

import PropTypes from 'prop-types';

/**
 * Defines the type alias for a `Contribution` record.
 */
export type ContributionRecord = {
  amount: number,
  legislatorId: string,
  organization: string,
};

/**
 * Defines the `PropTypes` validation structure for a `Contribution` record.
 */
export const PropType = PropTypes.shape({
  amount: PropTypes.number.isRequired,
  legislatorId: PropTypes.string.isRequired,
  organization: PropTypes.string.isRequired,
});
