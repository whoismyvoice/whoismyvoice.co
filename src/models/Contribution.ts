/**
 * Defines the type alias for a `Contribution` record.
 */
export interface Contribution {
  amount: number;
  legislatorId: string;
  organization: string;
}
