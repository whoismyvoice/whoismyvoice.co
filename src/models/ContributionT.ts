/**
 * Defines the type alias for a `Contribution` record.
 */
export type Record = {
  amount: number;
  legislatorId: string;
  organization: string;
};
