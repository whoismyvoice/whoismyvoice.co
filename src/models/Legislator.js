// @flow

import PropTypes from 'prop-types';
import { PropType as ContributionType } from './Contribution';

// Types
import type { ContributionRecord } from './Contribution';

export type BioguideId = string;
export type FecId = string;
opaque type LegislatorChamber = 'senate' | 'house' | 'unknown';
export type LegislatorIdentifier = string;

type LegislatorBioRecord = {
  birthday: string,
  gender: string,
};

export type LegislatorChannelRecord = {
  id: string,
  type: string,
};

type LegislatorIdRecord = {
  bioguide: BioguideId,
  fec: Array<FecId>,
};

type LegislatorNameRecord = {
  first?: string,
  last?: string,
  official_full: string,
};

type LegislatorTermRecord = {
  type: string,
  start: string,
  end: string,
  state: string,
  district?: string,
  party: string,
  phone: string,
};

export type LegislatorRecord = {
  id: LegislatorIdRecord,
  identifier?: string,
  name: string | LegislatorNameRecord,
  bio: LegislatorBioRecord,
  channels: Array<LegislatorChannelRecord>,
  photoUrl?: string,
  terms: Array<LegislatorTermRecord>,
};

/**
 * Defines the `PropTypes` validation structure for a `Legislator` record.
 */
export const PropType = PropTypes.shape({
  id: PropTypes.shape({
    bioguide: PropTypes.string,
    fec: PropTypes.arrayOf(PropTypes.string),
  }),
  identifier: PropTypes.string,
  name: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.shape({
      first: PropTypes.string,
      last: PropTypes.string,
      official_full: PropTypes.string.isRequired,
    }),
  ]),
  bio: PropTypes.shape({
    birthday: PropTypes.string,
    gender: PropTypes.string,
  }),
  channels: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      type: PropTypes.string,
    })
  ),
  photoUrl: PropTypes.string,
  terms: PropTypes.arrayOf(
    PropTypes.shape({
      type: PropTypes.string.isRequired,
      start: PropTypes.string.isRequired,
      end: PropTypes.string.isRequired,
      state: PropTypes.string,
      district: PropTypes.number,
      party: PropTypes.string,
    })
  ),
});

/**
 * Remove accents and diacritic marks from the given string.
 * @param {string} str to clone without 'special' characters.
 * @returns `str` with the accents and diacritic characters replaced with their
 *    plain latin variants.
 */
function stripDiacritics(str) {
  return str.normalize('NFD').replace(/[\u0300-\u036f]/g, '');
}

/**
 * Provides some convenience methods on top of the record for a legislator.
 */
export class Legislator {
  /**
   * Get the legislator's identifier from the given `record`.
   * @param {LegislatorRecord} record that matches the `PropType` defined for
   *    `Legislator` instances.
   * @returns the identifier used for legislator records for the given `record`.
   */
  static getIdentifier(record: LegislatorRecord): LegislatorIdentifier {
    PropTypes.checkPropTypes(
      { record: PropType },
      { record },
      'record',
      'Legislator#getIdentifier'
    );
    let identifier = undefined;
    if (record.identifier !== undefined) {
      identifier = record.identifier;
    } else if (typeof record.name === 'string') {
      identifier = stripDiacritics(record.name);
    } else {
      identifier = stripDiacritics(record.name.official_full);
    }
    return identifier;
  }

  /**
   * Retrieve the first contribution out of `contributions` that matches the identifier
   * of the given legislator.
   * @param {Array<ContributionRecord>} contributions records to be searched.
   * @param {LegislatorRecord|Legislator} legislator instance or record for which contributions
   *    will be searched.
   * @returns `-1` if no records from `contributions` match `legislator`, `amount`
   *    property of the contribution record otherwise.
   */
  static getContributionAmount(
    contributions: Array<ContributionRecord>,
    legislator: LegislatorRecord | Legislator
  ): number {
    PropTypes.checkPropTypes(
      { contributions: PropTypes.arrayOf(ContributionType) },
      { contributions },
      'contributions',
      'Legislator#getContributionAmount'
    );
    const id =
      legislator instanceof Legislator
        ? legislator.identifier
        : Legislator.getIdentifier(legislator);
    const contribution = contributions.find(
      record => record.legislatorId === id
    );
    if (contribution !== undefined) {
      return contribution.amount;
    } else {
      return -1;
    }
  }

  record: LegislatorRecord;

  /**
   * Create a new Legislator instance from the data provided in the given
   * `record`.
   *
   *      {
   *        "terms": [
   *          {
   *            "type": "rep|sen",
   *            "start": "<date string>",
   *            "end": "<date string>",
   *            "state": "<US State abbreviation>",
   *            "party": "<US Political Party Name>",
   *          }
   *        ]
   *      }
   *
   * @param {LegislatorRecord} record containing data about the legislator.
   */
  constructor(record: LegislatorRecord) {
    PropTypes.checkPropTypes(
      { record: PropType },
      { record },
      'record',
      'Legislator'
    );
    this.record = record;
  }

  /**
   * Gets the biographical object for the instance.
   * @returns an object containing (at least) `birthday` and `gender` properties.
   */
  get bio(): LegislatorBioRecord {
    return this.record.bio;
  }

  /**
   * Retrieve the bioguide property out of `record.id`.
   * @returns the bioguide id assigned to this legislator.
   */
  get bioguide(): BioguideId {
    return this.id.bioguide;
  }

  /**
   * Get the chamber of the legislator.
   * @returns 'house' if the legislator's current term is in the US House of
   *    Representatives; 'senate' if the legislator's current term is in the
   *    US Senate; 'unknown' otherwise.
   */
  get chamber(): LegislatorChamber {
    if (this.isSenator) {
      return 'senate';
    } else if (this.isRepresentative) {
      return 'house';
    } else {
      return 'unknown';
    }
  }

  /**
   * Retrieve the most recent term object from the given record.
   * @returns a term record for the most recent (current) term.
   */
  get currentTerm(): LegislatorTermRecord {
    const { terms } = this;
    const [currentTerm] = terms.slice(-1);
    return currentTerm;
  }

  /**
   * Retrieve the full, official name of the Legislator.
   * @returns full, official name.
   */
  get fullName(): string {
    return this.name.official_full;
  }

  /**
   * Retrieve the abbreviated gender of this legislator.
   * @returns the `gender` recorded for the legislator.
   */
  get gender(): string {
    return this.bio.gender;
  }

  /**
   * Retrieve a pronoun appropriate to the `gender` property.
   * @returns 'Them' if `gender` is anything other than 'M' or 'F'.
   */
  get genderPronoun(): string {
    switch (this.gender) {
      case 'M':
        return 'Him';
      case 'F':
        return 'Her';
      default:
        return 'Them';
    }
  }

  /**
   * Gets the id object for the instance.
   * @returns an object containing (at least) `fec` and `bioguide` properties.
   */
  get id(): LegislatorIdRecord {
    return this.record.id;
  }

  /**
   * Uses `Legislator#getIdentifier(record)` using the record provided at
   * instance creation time.
   * @returns the identifier used for legislator records for the given `record`.
   */
  get identifier(): LegislatorIdentifier {
    return Legislator.getIdentifier(this.record);
  }

  /**
   * Whether or not the `currentTerm` reflects a US Senate term.
   * @returns `true` if the `currentTerm` exists and has a Senate type.
   */
  get isSenator(): boolean {
    return this.currentTerm !== undefined && this.currentTerm.type === 'sen';
  }

  /**
   * Whether or not the `currentTerm` reflects a US House of Representatives
   * term.
   * @returns `true` if the `currentTerm` exists and has a US House of
   *    Representatives type.
   */
  get isRepresentative(): boolean {
    return this.currentTerm !== undefined && this.currentTerm.type === 'rep';
  }

  /**
   * Gets the name object for the instance.
   * @returns an object containing (at least) the `official_full` property.
   */
  get name(): LegislatorNameRecord {
    if (typeof this.record.name === 'string') {
      return { official_full: this.record.name };
    } else {
      return this.record.name;
    }
  }

  /**
   * Retrieves the US Political Party under which this legislator was elected
   * for their current term.
   * @returns current US Political Party.
   */
  get party(): string {
    return this.currentTerm.party;
  }

  /**
   * Get the phone number associated with the current term.
   * @returns The phone number associated with the current term.
   */
  get phone(): string {
    return this.currentTerm.phone;
  }

  /**
   * Get the URL for the official bioguide photo.
   * @returns a URL on theunitedstates.io.
   */
  get photoUrl(): string {
    const bioguide = this.bioguide;
    const size = 'original';
    return `https://theunitedstates.io/images/congress/${size}/${bioguide}.jpg`;
  }

  /**
   * The US State abbreviation for the state this legislator represents.
   * @returns a US State abbreviation.
   */
  get state(): string {
    return this.currentTerm.state;
  }

  /**
   * Gets the array of terms for the instance.
   * @returns an array of term records containing (at least) one record.
   */
  get terms(): Array<LegislatorTermRecord> {
    return this.record.terms;
  }

  /**
   * Get the Twitter handle of the legislator if it is provided.
   * @returns `undefined` if it isn't provided in `record`; the Twitter handle
   *    if it can be found in `record`.
   */
  get twitter(): ?string {
    const channels = this.record.channels || [];
    const channel = channels.find(channel => channel.type === 'Twitter');
    return channel === undefined ? undefined : channel.id;
  }
}

export default Legislator;
