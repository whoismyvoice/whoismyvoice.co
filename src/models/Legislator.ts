import { Contribution } from './Contribution';
import { Record as Official } from './Official';

export type TermType = 'sen' | 'rep';
export enum Party {
  Democrat = 'Democrat',
  Republican = 'Republican',
  Independent = 'Independent',
}
export type BioguideId = string;
export type FecId = string;
export enum Chamber {
  Senate,
  House,
  Unknown,
}
export type Identifier = string;

export interface BioRecord {
  birthday: string;
  gender: string;
}

export interface ChannelRecord {
  id: string;
  type: string;
}

export interface IdRecord {
  bioguide: BioguideId;
  fec: Array<FecId>;
}

export interface NameRecord {
  first?: string;
  last?: string;
  official_full: string;
}

export interface TermRecord {
  type: TermType;
  start: string;
  end: string;
  state: string;
  district?: string;
  party: Party;
  phone: string;
}

export interface Record {
  id: IdRecord;
  identifier?: string;
  name: string | NameRecord;
  bio: BioRecord;
  channels: Array<ChannelRecord>;
  photoUrl?: string;
  terms: Array<TermRecord>;
}

/**
 * Remove accents and diacritic marks from the given string.
 * @param {string} str to clone without 'special' characters.
 * @returns `str` with the accents and diacritic characters replaced with their
 *    plain latin variants.
 */
function stripDiacritics(str: string) {
  return str.normalize('NFD').replace(/[\u0300-\u036f]/g, '');
}

/**
 * Provides some convenience methods on top of the record for a legislator.
 */
export class Legislator implements Record {
  /**
   * Holds the Record used to retrieve data for getters.
   */
  private record: Record;

  /**
   * Get the legislator's identifier from the given `record`.
   * @param {Record | Official | Legislator} record that matches the `PropType`
   *    defined for `Legislator` instances.
   * @returns the identifier used for legislator records for the given `record`.
   */
  static getIdentifier(record: Record | Official | Legislator): Identifier {
    let identifier = undefined;
    if (record instanceof Legislator) {
      identifier = record.identifier;
    } else if (typeof record.name === 'string') {
      identifier = stripDiacritics(record.name);
    } else if (typeof record.name !== 'undefined') {
      identifier = stripDiacritics(record.name.official_full);
    } else {
      identifier = record.identifier || 'unknown';
    }
    return identifier;
  }

  /**
   * Retrieve the first contribution out of `contributions` that matches the identifier
   * of the given legislator.
   * @param {Array<ContributionRecord>} contributions records to be searched.
   * @param {Record|Legislator} legislator instance or record for which contributions
   *    will be searched.
   * @returns `-1` if no records from `contributions` match `legislator`, `amount`
   *    property of the contribution record otherwise.
   */
  static getContributionAmount(
    contributions: Array<Contribution>,
    legislator: Record | Legislator
  ): number {
    const id =
      legislator instanceof Legislator
        ? legislator.identifier
        : Legislator.getIdentifier(legislator);
    const contribution = contributions.find(
      record => record.legislatorId === id
    );
    if (contribution !== undefined) {
      return contribution.amount;
    } else {
      return -1;
    }
  }

  /**
   * Create a new Legislator instance from the data provided in the given
   * `record`.
   *
   *      {
   *        "terms": [
   *          {
   *            "type": "rep|sen",
   *            "start": "<date string>",
   *            "end": "<date string>",
   *            "state": "<US State abbreviation>",
   *            "party": "<US Political Party Name>",
   *          }
   *        ]
   *      }
   *
   * @param {Record} record containing data about the legislator.
   */
  constructor(record: Record) {
    this.record = record;
  }

  /**
   * Gets the biographical object for the instance.
   * @returns an object containing (at least) `birthday` and `gender` properties.
   */
  get bio(): BioRecord {
    return this.record.bio;
  }

  /**
   * Retrieve the bioguide property out of `record.id`.
   * @returns the bioguide id assigned to this legislator.
   */
  get bioguide(): BioguideId {
    return this.id.bioguide;
  }

  /**
   * Get the chamber of the legislator.
   * @returns 'house' if the legislator's current term is in the US House of
   *    Representatives; 'senate' if the legislator's current term is in the
   *    US Senate; 'unknown' otherwise.
   */
  get chamber(): Chamber {
    if (this.isSenator) {
      return Chamber.Senate;
    } else if (this.isRepresentative) {
      return Chamber.House;
    } else {
      return Chamber.Unknown;
    }
  }

  /**
   * Retrieve the channels from by which the legislator can be reached.
   * @returns an array of channels.
   */
  get channels(): Array<ChannelRecord> {
    return this.record.channels;
  }

  /**
   * Retrieve the most recent term object from the given record.
   * @returns a term record for the most recent (current) term.
   */
  get currentTerm(): TermRecord {
    const { terms } = this;
    const [currentTerm] = terms.slice(-1);
    return currentTerm;
  }

  /**
   * Retrieve the full, official name of the Legislator.
   * @returns full, official name.
   */
  get fullName(): string {
    return this.name.official_full;
  }

  /**
   * Retrieve the abbreviated gender of this legislator.
   * @returns the `gender` recorded for the legislator.
   */
  get gender(): string {
    return this.bio.gender;
  }

  /**
   * Retrieve a pronoun appropriate to the `gender` property.
   * @returns 'Them' if `gender` is anything other than 'M' or 'F'.
   */
  get genderPronoun(): string {
    switch (this.gender) {
      case 'M':
        return 'Him';
      case 'F':
        return 'Her';
      default:
        return 'Them';
    }
  }

  /**
   * Gets the id object for the instance.
   * @returns an object containing (at least) `fec` and `bioguide` properties.
   */
  get id(): IdRecord {
    return this.record.id;
  }

  /**
   * Uses `Legislator#getIdentifier(record)` using the record provided at
   * instance creation time.
   * @returns the identifier used for legislator records for the given `record`.
   */
  get identifier(): Identifier {
    return Legislator.getIdentifier(this.record);
  }

  /**
   * Whether or not the `currentTerm` reflects a US Senate term.
   * @returns `true` if the `currentTerm` exists and has a Senate type.
   */
  get isSenator(): boolean {
    return this.currentTerm !== undefined && this.currentTerm.type === 'sen';
  }

  /**
   * Whether or not the `currentTerm` reflects a US House of Representatives
   * term.
   * @returns `true` if the `currentTerm` exists and has a US House of
   *    Representatives type.
   */
  get isRepresentative(): boolean {
    return this.currentTerm !== undefined && this.currentTerm.type === 'rep';
  }

  /**
   * Gets the name object for the instance.
   * @returns an object containing (at least) the `official_full` property.
   */
  get name(): NameRecord {
    if (typeof this.record.name === 'string') {
      return { official_full: this.record.name };
    } else {
      return this.record.name;
    }
  }

  /**
   * Retrieves the US Political Party under which this legislator was elected
   * for their current term.
   * @returns current US Political Party.
   */
  get party(): Party {
    return this.currentTerm.party;
  }

  /**
   * Get the phone number associated with the current term.
   * @returns The phone number associated with the current term.
   */
  get phone(): string {
    return this.currentTerm.phone;
  }

  /**
   * Get the URL for the official bioguide photo.
   * @returns a URL on theunitedstates.io.
   */
  get photoUrl(): string {
    const bioguide = this.bioguide;
    const size = 'original';
    return `https://theunitedstates.io/images/congress/${size}/${bioguide}.jpg`;
  }

  /**
   * The US State abbreviation for the state this legislator represents.
   * @returns a US State abbreviation.
   */
  get state(): string {
    return this.currentTerm.state;
  }

  /**
   * Gets the array of terms for the instance.
   * @returns an array of term records containing (at least) one record.
   */
  get terms(): Array<TermRecord> {
    return this.record.terms;
  }

  /**
   * Get the Twitter handle of the legislator if it is provided.
   * @returns `undefined` if it isn't provided in `record`; the Twitter handle
   *    if it can be found in `record`.
   */
  get twitter(): string | undefined {
    const channels = this.record.channels || [];
    const channel = channels.find(c => c.type === 'Twitter');
    return channel === undefined ? undefined : channel.id;
  }
}

export default Legislator;
