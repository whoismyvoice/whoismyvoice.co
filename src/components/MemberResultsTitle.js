import PropTypes from 'prop-types';

import { Legislator, } from '../models/Legislator';
import { Title } from './Title';

import '../styles/TitleComponent.css';

export class MemberResultsTitle extends Title {
  static defaultProps = {
    legislators: [],
  }

  static propTypes = {
    legislators: PropTypes.arrayOf(PropTypes.instanceOf(Legislator)),
  }

  constructor(props) {
    super(props);
    this.state = {
      templateData: {
        ...(props.templateData),
        memberType: MemberResultsTitle.getMemberType(props.legislators),
      },
    };
  }

  componentWillReceiveProps(nextProps, nextContext) {
    const { legislators, } = nextProps;
    this.setState({
      templateData: {
        ...(this.state.templateData),
        ...(nextProps.templateData),
        memberType: MemberResultsTitle.getMemberType(legislators),
      },
    });
  }

  static getMemberType(legislators) {
    let memberType;
    if (legislators.length === 1) {
      const [ rep, ] = legislators;
      memberType = rep.chamber === 'senate' ? 'Senator' : 'Representative';
    } else if (legislators.length === 2) {
      const chambers = legislators.map(legislator => legislator.chamber);
      const houseIndex = chambers.findIndex(chamber => chamber === 'house');
      if (houseIndex === -1) {
        memberType = 'Senators';
      } else if (houseIndex === 0) {
        memberType = 'Representative & Senator';
      } else if (houseIndex === 1) {
        memberType = 'Senator & Representative';
      }
    } else if (legislators.length === 3) {
      memberType = 'Senators & Representative';
    } else {
      throw new Error('Too many legislators provided.');
    }
    return memberType;
  }
}
