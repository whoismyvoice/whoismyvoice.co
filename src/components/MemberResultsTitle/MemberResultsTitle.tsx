import { Chamber, Legislator } from '../../models/Legislator';
import {
  Props as TitleProps,
  State as TitleState,
  Title,
} from '../Title/Title';

import '../../styles/TitleComponent.scss';

interface Props extends TitleProps {
  legislators: Array<Legislator>;
}

type State = TitleState;

export function getMemberType(legislators: Array<Legislator>) {
  let memberType;
  const { Senate, House } = Chamber;
  if (legislators.length === 1) {
    const [rep] = legislators;
    return rep.chamber === Senate ? 'Senator' : 'Representative';
  } else if (legislators.length === 2) {
    const chambers = legislators.map(legislator => legislator.chamber);
    const houseIndex = chambers.findIndex(chamber => chamber === House);
    if (houseIndex === -1) {
      memberType = 'Senators';
    } else if (houseIndex === 0) {
      memberType = 'Representative & Senator';
    } else {
      // so `houseIndex === 1`
      memberType = 'Senator & Representative';
    }
  } else if (legislators.length === 3) {
    memberType = 'Senators & Representative';
  } else if (legislators.length === 0) {
    throw new Error('Not enough legislators provided.');
  } else {
    throw new Error('Too many legislators provided.');
  }
  return memberType;
}

export class MemberResultsTitle extends Title<Props, State> {
  static defaultProps = {
    ...Title.defaultProps,
    legislators: [],
  };

  constructor(props: Props) {
    super(props);
    this.state = {
      templateData: {
        ...props.templateData,
        memberType: getMemberType(props.legislators),
      },
    };
  }

  UNSAFE_componentWillReceiveProps(nextProps: Readonly<Props>) {
    const { legislators } = nextProps;
    this.setState({
      templateData: {
        ...this.state.templateData,
        ...nextProps.templateData,
        memberType: getMemberType(legislators),
      },
    });
  }
}
