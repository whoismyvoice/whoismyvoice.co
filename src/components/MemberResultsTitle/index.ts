import { MemberResultsTitle as Component } from './MemberResultsTitle';

export const MemberResultsTitle = Component;
export default Component;
