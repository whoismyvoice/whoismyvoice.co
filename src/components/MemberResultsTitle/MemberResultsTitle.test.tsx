import React from 'react';
import { render } from '@testing-library/react';
import { MemberResultsTitle, getMemberType } from './MemberResultsTitle';
import { createLegislator, createTerm } from '../../models/Legislator.test';
import { Legislator, TermRecord } from '../../models/Legislator';

function createLegislatorWithTerm(
  name: string,
  terms: Array<TermRecord>
): Legislator {
  const record = createLegislator(name);
  return new Legislator({
    ...record,
    terms,
  });
}

describe(getMemberType, () => {
  describe('empty legislators', () => {
    const legislators: Array<Legislator> = [];

    it('throws an error', () => {
      expect(getMemberType.bind(legislators, legislators)).toThrow(Error);
    });
  });

  describe('single legislator', () => {
    it('is senator', () => {
      const legislators = [
        createLegislatorWithTerm('John Smith', [createTerm('sen')]),
      ];
      expect(getMemberType(legislators)).toBe('Senator');
    });

    it('is rep', () => {
      const legislators = [
        createLegislatorWithTerm('John Smith', [createTerm('rep')]),
      ];
      expect(getMemberType(legislators)).toBe('Representative');
    });
  });

  describe('two legeslators', () => {
    it('is senators', () => {
      const legislators = [
        createLegislatorWithTerm('John Smith', [createTerm('sen')]),
        createLegislatorWithTerm('John Smith Jr.', [createTerm('sen')]),
      ];
      expect(getMemberType(legislators)).toBe('Senators');
    });

    it('is senator & rep', () => {
      const legislators = [
        createLegislatorWithTerm('John Smith', [createTerm('sen')]),
        createLegislatorWithTerm('John Smith Jr.', [createTerm('rep')]),
      ];
      expect(getMemberType(legislators)).toBe('Senator & Representative');
    });

    it('is rep & senator', () => {
      const legislators = [
        createLegislatorWithTerm('John Smith', [createTerm('rep')]),
        createLegislatorWithTerm('John Smith Jr.', [createTerm('sen')]),
      ];
      expect(getMemberType(legislators)).toBe('Representative & Senator');
    });
  });

  describe('three legeslators', () => {
    it('is senators & rep', () => {
      const legislators = [
        createLegislatorWithTerm('John Smith', [createTerm('sen')]),
        createLegislatorWithTerm('John Smith Jr.', [createTerm('sen')]),
        createLegislatorWithTerm('John Smith III', [createTerm('rep')]),
      ];
      expect(getMemberType(legislators)).toBe('Senators & Representative');
    });
  });

  describe('more legeslators', () => {
    it('is senators & rep', () => {
      const legislators = [
        createLegislatorWithTerm('John Smith', [createTerm('sen')]),
        createLegislatorWithTerm('John Smith Jr.', [createTerm('sen')]),
        createLegislatorWithTerm('John Smith III', [createTerm('rep')]),
        createLegislatorWithTerm('John Smith IV', [createTerm('rep')]),
      ];
      expect(getMemberType.bind(legislators, legislators)).toThrow(Error);
    });
  });
});

describe('MemberResultsTitle', () => {
  describe('single legislator', () => {
    it('is senator', () => {
      const legislators = [
        createLegislatorWithTerm('John Smith', [createTerm('sen')]),
      ];
      const props = { legislators };
      const { getByText } = render(<MemberResultsTitle {...props} />);
      expect(getByText(getMemberType(legislators))).not.toBeNull();
    });

    it('is rep', () => {
      const legislators = [
        createLegislatorWithTerm('John Smith', [createTerm('rep')]),
      ];
      const props = { legislators };
      const { getByText } = render(<MemberResultsTitle {...props} />);
      expect(getByText(getMemberType(legislators))).not.toBeNull();
    });
  });

  describe('two legislators', () => {
    it('is senators', () => {
      const legislators = [
        createLegislatorWithTerm('John Smith', [createTerm('sen')]),
        createLegislatorWithTerm('John Smith Jr.', [createTerm('sen')]),
      ];
      const props = { legislators };
      const { getByText } = render(<MemberResultsTitle {...props} />);
      expect(getByText(getMemberType(legislators))).not.toBeNull();
    });

    it('is senator & rep', () => {
      const legislators = [
        createLegislatorWithTerm('John Smith', [createTerm('sen')]),
        createLegislatorWithTerm('John Smith Jr.', [createTerm('rep')]),
      ];
      const props = { legislators };
      const { getByText } = render(<MemberResultsTitle {...props} />);
      expect(getByText(getMemberType(legislators))).not.toBeNull();
    });

    it('is rep & senator', () => {
      const legislators = [
        createLegislatorWithTerm('John Smith', [createTerm('rep')]),
        createLegislatorWithTerm('John Smith Jr.', [createTerm('sen')]),
      ];
      const props = { legislators };
      const { getByText } = render(<MemberResultsTitle {...props} />);
      expect(getByText(getMemberType(legislators))).not.toBeNull();
    });
  });

  describe('three legeslators', () => {
    it('is senators & rep', () => {
      const legislators = [
        createLegislatorWithTerm('John Smith', [createTerm('sen')]),
        createLegislatorWithTerm('John Smith Jr.', [createTerm('sen')]),
        createLegislatorWithTerm('John Smith III', [createTerm('rep')]),
      ];
      const props = { legislators };
      const { getByText } = render(<MemberResultsTitle {...props} />);
      expect(getByText(getMemberType(legislators))).not.toBeNull();
    });
  });
});
