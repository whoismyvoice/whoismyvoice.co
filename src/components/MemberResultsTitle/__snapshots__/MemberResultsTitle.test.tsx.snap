// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MemberResultsTitle single legislator is rep 1`] = `
<div
  className="title-component "
>
  <div
    className="title-component__description"
    dangerouslySetInnerHTML={
      Object {
        "__html": "",
      }
    }
  />
</div>
`;

exports[`MemberResultsTitle single legislator is senator 1`] = `
<div
  className="title-component "
>
  <div
    className="title-component__description"
    dangerouslySetInnerHTML={
      Object {
        "__html": "",
      }
    }
  />
</div>
`;

exports[`MemberResultsTitle three legeslators is senators & rep 1`] = `
<div
  className="title-component "
>
  <div
    className="title-component__description"
    dangerouslySetInnerHTML={
      Object {
        "__html": "",
      }
    }
  />
</div>
`;

exports[`MemberResultsTitle two legislators is rep & senator 1`] = `
<div
  className="title-component "
>
  <div
    className="title-component__description"
    dangerouslySetInnerHTML={
      Object {
        "__html": "",
      }
    }
  />
</div>
`;

exports[`MemberResultsTitle two legislators is senator & rep 1`] = `
<div
  className="title-component "
>
  <div
    className="title-component__description"
    dangerouslySetInnerHTML={
      Object {
        "__html": "",
      }
    }
  />
</div>
`;

exports[`MemberResultsTitle two legislators is senators 1`] = `
<div
  className="title-component "
>
  <div
    className="title-component__description"
    dangerouslySetInnerHTML={
      Object {
        "__html": "",
      }
    }
  />
</div>
`;
